webapp/tests/karma/js/enketo/widgets/external-app-launcher.spec.ts

```diff
@@ -0,0 +1,619 @@
+                </label>
+                <label class="question">
+                  <input type="text" name="/main/app/inputs/repeat/addresses/house_info/house_number">
+                </label>
+              </section>
+            </section>
+            <section 
+            class="or-appearance-external-app-object-list or-group repeat-2" 
+            name="/main/app/inputs/repeat/addresses">
+              <label class="question">
+                <input type="text" name="/main/app/inputs/repeat/addresses/city">
+              </label>
+              <label class="question">
+                <input type="text" name="/main/app/inputs/repeat/addresses/country">
+              </label>
+              <section class="or-appearance-external-app-object or-group" name="/main/app/inputs/address/house_info">
+                <label class="question">
+                  <input type="text" name="/main/app/inputs/repeat/addresses/house_info/description">
+                </label>
+                <label class="question">
+                  <input type="text" name="/main/app/inputs/repeat/addresses/house_info/house_number">
+                </label>
+              </section>
+            </section>
+          </section>
+        </section>
+      `;
+
+      $widget
+        .append(appSettings)
+        .append(outputs);
+
+      $launchAppBtn.click();
+      await Promise.resolve();
+
+      expect(window.CHTCore.ExternalAppLauncher.launchExternalApp.callCount).to.equal(1);
+      expect(window.CHTCore.ExternalAppLauncher.launchExternalApp.args[0]).to.have.deep.members([{
+        action: 'org.example.action.VIEW'
+      }]);
+      expect($widget.find('input[name="/main/app/inputs/first_name"]').val()).to.equal('Jack');
+      expect($widget.find('input[name="/main/app/inputs/last_name"]').val()).to.equal('Doors');
+      expect($widget.find('.repeat-1 input[name="/main/app/inputs/repeat/addresses/city"]').val())
+        .to.equal('Hokkaido');
+      expect($widget.find('.repeat-1 input[name="/main/app/inputs/repeat/addresses/country"]').val())
+        .to.equal('Japan');
+      expect($widget.find('.repeat-1 input[name="/main/app/inputs/repeat/addresses/house_info/description"]').val())
+        .to.equal('House is a shell');
+      expect($widget.find('.repeat-1 input[name="/main/app/inputs/repeat/addresses/house_info/house_number"]').val())
+        .to.equal('15');
+      expect($widget.find('.repeat-2 input[name="/main/app/inputs/repeat/addresses/city"]').val())
+        .to.equal('Seattle');
+      expect($widget.find('.repeat-2 input[name="/main/app/inputs/repeat/addresses/country"]').val())
+        .to.equal('USA');
+      expect($widget.find('.repeat-2 input[name="/main/app/inputs/repeat/addresses/house_info/description"]').val())
+        .to.equal('House is a pineapple');
+      expect($widget.find('.repeat-2 input[name="/main/app/inputs/repeat/addresses/house_info/house_number"]').val())
+        .to.equal('5');
+    });
+  });
+});
```

Comment:
Great work on this test!

There is a whole lot of HTML inlined which is hard to read but also distracts from the test. It also makes it really hard to reuse outside of this spec. Could you please pull these out into a utility factory so that they can be reused? Potentially if you take in some parameters you'll find there's some reuse of the HTML that you can do already so you don't have to repeat the boilerplate over and over again.

Reply:
Done!